rust/processor/src/db/common/models/emojicoin_models/queries/leaderboard.sql

```diff
@@ -0,0 +1,113 @@
+-- Insert a snapshot of the leaderboard in arena_leaderboard_history for the
+-- given melee ID.
+
+-- The performance of this query has not been established, but should be
+-- irrelevant, as arenas end once every 24 hours, and this script only runs
+-- when an arena ends
+
+
+-- This query takes three parameters:
+--
+-- $1: the melee_id for which to calculate the leaderboard
+-- $2: the curve price of emojicoin_0 at the end of the melee
+-- $3: the curve price of emojicoin_1 at the end of the melee
+
+INSERT INTO arena_leaderboard_history (
+    "user",
+    melee_id,
+    profits,
+    losses,
+    emojicoin_0_balance,
+    emojicoin_1_balance,
+    exited,
+    last_exit_0,
+    withdrawals
+)
+-- Get the total deposit amount and last exit for each user.
+WITH position AS (
+    SELECT
+        "user",
+        deposits,
+        last_exit_0
+    FROM arena_position
+    WHERE melee_id = $1
+),
+-- Get the total withdrawal amount for each user.
+withdrawals AS (
+    SELECT
+        "user",
+        SUM(apt_proceeds) as withdrawals
+    FROM arena_exit_events
+    WHERE melee_id = $1
+    AND during_melee
+    GROUP BY "user"
+),
+-- Get the last balance the user had at the end of the melee.
+last_balances AS (
+    SELECT DISTINCT ON("user")
+        "user",
+        emojicoin_0_balance,
+        emojicoin_1_balance
+    FROM (
+        SELECT
+            "user",
+            transaction_version,
+            event_index,
+            emojicoin_0_proceeds AS emojicoin_0_balance,
+            emojicoin_1_proceeds AS emojicoin_1_balance
+        FROM arena_enter_events
+        WHERE melee_id = $1
+        UNION ALL
+        SELECT
+            "user",
+            transaction_version,
+            event_index,
+            emojicoin_0_proceeds AS emojicoin_0_balance,
+            emojicoin_1_proceeds AS emojicoin_1_balance
+        FROM arena_swap_events
+        WHERE melee_id = $1
+        AND during_melee
+        UNION ALL
+        SELECT
+            "user",
+            transaction_version,
+            event_index,
+            0::numeric AS emojicoin_0_balance,
+            0::numeric AS emojicoin_1_balance
+        FROM arena_exit_events
+        WHERE melee_id = $1
+        AND during_melee
+    ) AS a
+    ORDER BY "user", transaction_version DESC, event_index DESC
+)
+SELECT
+    position."user",
+    $1 AS melee_id,
+    -- Proifts = Withdrawals in APT + current emojicoin balance converted to APT.
+    COALESCE(withdrawals, 0) +
+        ROUND(
+            emojicoin_0_balance * $2 +
+            emojicoin_1_balance * $3
+        ) AS profits,
+    deposits AS losses,
+    emojicoin_0_balance,
+    emojicoin_1_balance,
+    -- If user has no balance at the end of the melee, or he has, but there is
+    -- an exit event that happened after the end of the melee, then set exited
+    -- to true, otherwise to false.
```

Comment:
```suggestion
    -- Determine whether the user has exited or not by checking that one of the
    -- following conditions is true:
    --   1. The user has no balance at the end of the melee.
    --   2. There is an exit event for the user *after* that melee ended.
```

Also, after rewriting this and understanding better what it's doing, I'm wondering if this is always strictly true...are the emojicoin balances solely coming from the processor or would transfers in and out of their account (unrelated to the melee) affect their perceived "exit" status here?

Reply:
How could they move their escrow balance ?

Is that even possible ?

Reply:
Nope! I didn't realize it was a MyLibertyCoinStore (Alex's favorite term for unseizable coins in a `store` resource field 😆 ). Only the contract can move anything in and out

Was just making sure, and sort of thinking out loud